### PR TITLE
Update `azurerm_linux|windows_virtual_machine_scale_set` - Fix #8422

### LINKED ISSUE #8422: JSON parse error on beta version of inline extension block for azure_linux_virtual_machine_scale_set resource.
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.12.25
+ provider.azurerm v2.26.0

### Affected Resource(s)

* `azurerm_linux_virtual_machine_scale_set`

### Terraform Configuration Files

```hcl
resource "azurerm_linux_virtual_machine_scale_set" "manager" {
  name                = "manager"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
  sku                 = "Standard_DS2_v2"
  instances           = 3

  lifecycle {
    ignore_changes = [ instances ]
  }

  upgrade_mode = "Manual"

  scale_in_policy = "OldestVM"

  terminate_notification {
    enabled                          = true
    timeout                          = "PT5M"
  }
  overprovision                      = false

  source_image_id                    = var.image_id

  os_disk {
    caching                         = "ReadWrite"
    storage_account_type            = "Standard_LRS"
  }

  data_disk {
      caching                   = "ReadWrite"
      create_option             = "FromImage"
      disk_size_gb              = 128
      lun                       = 0
      storage_account_type      = "Standard_LRS"
  }

  data_disk {
      caching                   = "ReadWrite"
      create_option             = "FromImage"
      disk_size_gb              = 256
      lun                       = 1
      storage_account_type      = "Standard_LRS"
  }
  computer_name_prefix            = "manager"

  admin_username                  = "localadmin"
  disable_password_authentication = true
  
  admin_ssh_key {
    username                        = "localadmin"
    public_key                      = var.admin_public_key
  }

  network_interface {
    name    = "primary"
    primary = true

    ip_configuration {
      name                                    = "manager-ipconfig"
      primary                                 = true
      subnet_id                               = azurerm_subnet.docker.id
    }
  }

  extension {
    name                              = "data_disk_encryption"
    publisher                         = "Microsoft.Azure.Security"
    type                              = "AzureDiskEncryptionForLinux"
    type_handler_version              = "1.1"
    auto_upgrade_minor_version        = false
    settings                          = jsonencode({
          "EncryptionOperation"       = "EnableEncryption"
          "KeyVaultResourceId"        = azurerm_key_vault.vault.id
          "KeyVaultURL"               = azurerm_key_vault.vault.vault_uri
          "VolumeType"                = "DATA"
    })
  }
}
```

### Debug Output

https://gist.github.com/brentonoloughlin/189c312e321baaa22d2c1e64c727e1fc

### Expected Behavior

The Virtual Machine Scale Set resource should have been provisioned with the Azure Disk Encryption extension enabled.

the excerpt from terraform plan shows:
```
      + extension {
          + auto_upgrade_minor_version = false
          + name                       = "data_disk_encryption"
          + publisher                  = "Microsoft.Azure.Security"
          + settings                   = jsonencode(
                {
                  + EncryptionOperation = "EnableEncryption"
                  + KeyVaultResourceId  = "/subscriptions/37d41301-c255-4f06-95b4-902a713aeb5e/resourceGroups/vnet/providers/Microsoft.KeyVault/vaults/roughscale-vault"
                  + KeyVaultURL         = "https://roughscale-vault.vault.azure.net/"
                  + VolumeType          = "DATA"
                }
            )
          + type                       = "AzureDiskEncryptionForLinux"
          + type_handler_version       = "1.1"
        }
```

The same inline extension block code works successfully for the `azurerm_virtual_machine_scale_set` resource (with the same terraform version and azurerm provider version). For the 'azurerm_virtual_machine_scale_set' resource, the settings parameter from the terraform plan is slightly different (not sure if this is relevant):

```
      + extension {
          + auto_upgrade_minor_version = false
          + name                       = "data_disk_encryption"
          + provision_after_extensions = []
          + publisher                  = "Microsoft.Azure.Security"
          + settings                   = (known after apply)
          + type                       = "AzureDiskEncryptionForLinux"
          + type_handler_version       = "1.1"
        }
```

### Actual Behavior

`ARM_PROVIDER_VMSS_EXTENSIONS_BETA=true terraform apply` results in:
```
Error: failed to parse JSON from `settings`: unexpected end of JSON input

  on 60-vmss2.tf line 1, in resource "azurerm_linux_virtual_machine_scale_set" "manager":
   1: resource "azurerm_linux_virtual_machine_scale_set" "manager" {
```

### Steps to Reproduce

1. Build custom image with attached data disks (image built by packer based on RHEL7 Marketplace image)
2. `ARM_PROVIDER_VMSS_EXTENSIONS_BETA=true terraform apply`

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes #8422

### CODE REVIEW

**Comment by ArcturusZhang** [comment]
New test case is passing:
```
=== RUN   TestAccAzureRMLinuxVirtualMachineScaleSet_otherVmExtensionsOnlySettings
=== PAUSE TestAccAzureRMLinuxVirtualMachineScaleSet_otherVmExtensionsOnlySettings
=== CONT  TestAccAzureRMLinuxVirtualMachineScaleSet_otherVmExtensionsOnlySettings
--- PASS: TestAccAzureRMLinuxVirtualMachineScaleSet_otherVmExtensionsOnlySettings (359.83s)
PASS
```

File: azurerm/internal/services/compute/virtual_machine_scale_set.go [hunk 1]
```diff
@@ -1392,19 +1392,24 @@ func expandVirtualMachineScaleSetExtensions(input []interface{}) (*compute.Virtu
 			extensionProps.ForceUpdateTag = utils.String(forceUpdateTag.(string))
 		}
 
-		settings, err := structure.ExpandJsonFromString(extensionRaw["settings"].(string))
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse JSON from `settings`: %+v", err)
+		settingsRaw, ok := extensionRaw["settings"]
+		if ok && settingsRaw != "" {
+			settings, err := structure.ExpandJsonFromString(settingsRaw.(string))
```
Comment (jackofallops) [inline azurerm/internal/services/compute/virtual_machine_scale_set.go:1397]:
```suggestion
		settingsRaw, ok := extensionRaw["settings"].(string)
		if ok && settingsRaw != "" {
			settings, err := structure.ExpandJsonFromString(settingsRaw)
```

File: azurerm/internal/services/compute/virtual_machine_scale_set.go [hunk 2]
```diff
@@ -1392,19 +1392,24 @@ func expandVirtualMachineScaleSetExtensions(input []interface{}) (*compute.Virtu
...
+			extensionProps.Settings = settings
 		}
-		extensionProps.Settings = settings
 
-		protectedSettings, err := structure.ExpandJsonFromString(extensionRaw["protected_settings"].(string))
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse JSON from `settings`: %+v", err)
+		protectedSettingsRaw, ok := extensionRaw["protected_settings"]
+		if ok && protectedSettingsRaw != "" {
+			protectedSettings, err := structure.ExpandJsonFromString(protectedSettingsRaw.(string))
```
Comment (jackofallops) [inline azurerm/internal/services/compute/virtual_machine_scale_set.go:1406]:
```suggestion
		protectedSettingsRaw, ok := extensionRaw["protected_settings"].(string)
		if ok && protectedSettingsRaw != "" {
			protectedSettings, err := structure.ExpandJsonFromString(protectedSettingsRaw)
```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @ArcturusZhang 
Thanks for this, just a couple of really minor changes but otherwise LGTM 👍

File: azurerm/internal/services/compute/virtual_machine_scale_set.go (hunk 1 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/virtual_machine_scale_set.go:1397]: Done

File: azurerm/internal/services/compute/virtual_machine_scale_set.go (hunk 2 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/virtual_machine_scale_set.go]: Done

**Comment by ArcturusZhang** [comment]
Hi @jackofallops thanks for the review, I have made some changes please have a look, thanks

**Comment by ArcturusZhang** [comment]
And @jackofallops rather than doing this patch on this particular resource, maybe an update to the `structure.ExpandJsonFromString` to make it could accept an empty string is a better idea?

We have almost exactly the same issue here: https://github.com/terraform-providers/terraform-provider-azurerm/pull/8668

**Comment by jackofallops** [comment]
I had considered this, but changing that behaviour is likely to have unexpected effects elsewhere (such as potentially sending empty values where `nil` is expected). I think the current approach is appropriate for the time being, and is within our immediate domain of control, rather than changing the plugin-sdk.

**Review by jackofallops — APPROVED** [review]
Thanks @ArcturusZhang - LGTM now 👍

**Pull request merged**